usr/lib/common/asn1.c

```diff
@@ -411,8 +411,109 @@ CK_RV ber_decode_OCTET_STRING(CK_BYTE *str,
     return CKR_FUNCTION_FAILED;
 }
 
-//
-//
+/**
+ * Here we assume that the indication about unused bits is NOT
+ * part of the data. It will be added by this function.
+ */
+CK_ULONG ber_encode_BIT_STRING(CK_BBOOL length_only,
+                            CK_BYTE **ber_str,
+                            CK_ULONG *ber_str_len, CK_BYTE *data,
+                            CK_ULONG data_len,
+                            CK_BYTE unused_bits)
+{
+    CK_BYTE *buf = NULL;
+    CK_ULONG len;
+
+    // if data_len < 127 use short-form length id
+    // if data_len < 256 use long-form length id with 1-byte length field
+    // if data_len < 65536 use long-form length id with 2-byte length field
+    // if data_len < 16777216 use long-form length id with 3-byte length field
+
+    if (data_len + 1 < 128) {
+        len = 1 + 1 + 1 + data_len;
+    } else if (data_len + 1 < 256) {
+        len = 1 + (1 + 1) + 1 + data_len;
+    } else if (data_len + 1 < (1 << 16)) {
+        len = 1 + (1 + 2) + 1 + data_len;
+    } else if (data_len + 1 < (1 << 24)) {
+        len = 1 + (1 + 3) + 1 + data_len;
+    } else {
+        TRACE_ERROR("%s\n", ock_err(ERR_FUNCTION_FAILED));
+        return CKR_FUNCTION_FAILED;
+    }
+    if (length_only == TRUE) {
+        *ber_str_len = len;
```

Comment:
I know we are actually not doing this right now, but it think we sould use assertions at function entry to document calling conventions. E.g., this function must not be called with a NULL ber_str_len arg but you have to read the source code up to the first dereference to actually see it.

Reply:
see below ..